controllers/reclaimspacecronjob_controller.go

```diff
@@ -0,0 +1,370 @@
+	"k8s.io/apimachinery/pkg/runtime"
+	ref "k8s.io/client-go/tools/reference"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+// ReclaimSpaceCronJobReconciler reconciles a ReclaimSpaceCronJob object
+type ReclaimSpaceCronJobReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+}
+
+const (
+	jobOwnerKey = ".metadata.controller"
+	// default values for Spec parameters.
+	defaultFailedJobsHistoryLimit     int32 = 1
+	defaultSuccessfulJobsHistoryLimit int32 = 3
+)
+
+var (
+	apiGVStr                = csiaddonsv1alpha1.GroupVersion.String()
+	scheduledTimeAnnotation = csiaddonsv1alpha1.GroupVersion.Group + "/scheduled-at"
+)
+
+//+kubebuilder:rbac:groups=csiaddons.openshift.io,resources=reclaimspacecronjobs,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=csiaddons.openshift.io,resources=reclaimspacecronjobs/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=csiaddons.openshift.io,resources=reclaimspacecronjobs/finalizers,verbs=update
+//+kubebuilder:rbac:groups=csiaddons.openshift.io,resources=reclaimspacejobs,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=csiaddons.openshift.io,resources=reclaimspacejobs/status,verbs=get
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+func (r *ReclaimSpaceCronJobReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	_ = log.FromContext(ctx)
+	logger := log.FromContext(ctx)
+
+	// Fetch ReclaimSpaceCronJob instance
+	rsCronJob := &csiaddonsv1alpha1.ReclaimSpaceCronJob{}
+	err := r.Client.Get(ctx, req.NamespacedName, rsCronJob)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			logger.Info("ReclaimSpaceCronJob resource not found")
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	// set history limit defaults, if not specified.
+	if rsCronJob.Spec.FailedJobsHistoryLimit == nil {
+		*rsCronJob.Spec.FailedJobsHistoryLimit = defaultFailedJobsHistoryLimit
+	}
+	if rsCronJob.Spec.SuccessfulJobsHistoryLimit == nil {
+		*rsCronJob.Spec.SuccessfulJobsHistoryLimit = defaultSuccessfulJobsHistoryLimit
+	}
+
+	var childJobs csiaddonsv1alpha1.ReclaimSpaceJobList
+	if err := r.List(ctx, &childJobs, client.InNamespace(req.Namespace), client.MatchingFields{jobOwnerKey: req.Name}); err != nil {
+		logger.Error(err, "Failed to list child ReclaimSpaceJobs")
+		return ctrl.Result{}, err
+	}
+
+	// find the active,failed and successful list of jobs
+	var activeJob *csiaddonsv1alpha1.ReclaimSpaceJob
+	var successfulJobs, failedJobs []*csiaddonsv1alpha1.ReclaimSpaceJob
+	var mostRecentTime, lastSuccessfulTime *time.Time // find the last run so we can update the status
+
+	for i, job := range childJobs.Items {
```

Comment:
move this to a helper function and return required objects?

Reply:
> move this to a helper function and return required objects?

It would need to return 5 variables, IMO better to keep it inline in the same function flow.

Reply:
as we make more changes this becomes difficult to maintain. if you are concerned with returns variables create a  struct and return it

Reply:
> as we make more changes this becomes difficult to maintain. if you are concerned with returns variables create a struct and return it

done,ptal